### PR TITLE
Add deserializer impl

### PR DESCRIPTION
**Pull request by chmp**
Todo:

- [x] Improve API docs
- [x] Add API doc tests
- [x] Decide whether to implement Serializer or whether the ArrayBuilders are sufficient
  - Move into separate issue

### CODE REVIEW

File: serde_arrow/src/internal/deserialization/deserializer.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+use crate::internal::{
+    deserialization::outer_sequence_deserializer::OuterSequenceDeserializer,
+    error::{fail, Error, Result},
+};
+
+/// A structure to deserialize Arrow arrays into Rust objects
+///
+/// It can be constructed via
+///
```
Comment (ryzhyk) [inline serde_arrow/src/internal/deserialization/deserializer.rs:11]: Incomplete comment?

File: serde_arrow/src/internal/deserialization/deserializer.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+#[cfg_attr(has_arrow, doc = r"- [`Deserializer::from_record_batch`]")]
+#[cfg_attr(has_arrow, doc = r"- [`Deserializer::from_arrow`]")]
+#[cfg_attr(has_arrow2, doc = r"- [`Deserializer::from_arrow2`]")]
+///
+pub struct Deserializer<'de>(pub(crate) OuterSequenceDeserializer<'de>);
```
Comment (ryzhyk) [inline serde_arrow/src/internal/deserialization/deserializer.rs:16]: Should this struct still live in `internal`, given that it's now part of the public API?

**Review by ryzhyk — APPROVED** [review]
This matches our use case perfectly, thanks!

File: serde_arrow/src/internal/deserialization/deserializer.rs (hunk 1 above)
Comment (chmp) [reply serde_arrow/src/internal/deserialization/deserializer.rs:11]: Only, if no features are active :). Once you have arrow or arrow2 configured the bullet list below is included, but I will also make this part of the comment conditional.

File: serde_arrow/src/internal/deserialization/deserializer.rs (hunk 2 above)
Comment (chmp) [reply serde_arrow/src/internal/deserialization/deserializer.rs:16]: TBH. Also not happy with that. In particular as it is quite deep in the file tree. There is however precedence with the schema functionality. I will move it to the top level of the internal crate as a compromise :)

**Comment by chmp** [comment]
Thanks for the feedback!

**Pull request merged**